### ISSUE TITLE
Minju: Workflows, 2 weeks

### ISSUE THREAD

**Issue by minjupgeorge** [lab-brussels-1/home#168]
### 🥚 _[0. Local Development Without Git](./0-local-development-without-git)_
<hr/>

Practice the foundational workflows of software development by learning to write
Markdown locally on your own computer using Visual Studio Code (VSCode), the
Command Line Interface (CLI), and NPM scripts to automate your code's quality
(formatting, linting and spell checking).

- [x] 🥚 **Folder Structures**: You can explain how files and folders are stored
      in your computer and can find, open or create files in your computer
      without using the Command Line Interface.
- 🥚 **Command Line Interface (CLI)**: In a Unix shell you can ...
  - [x] open a new terminal window
  - [x] navigate up and down directories using `cd`
  - [x] list the contents of a directory using `ls`
  - [x] view the contents of a file using `cat`
  - [x] create new files using `touch`
  - [x] create new folders using `mkdir`
- [x] 🥚 **Markdown**: You can write a document in Markdown with no syntax
      mistakes that renders into a well-formatted document.
- 🥚 **VSCode**: You can complete these workflows in VScode, and can use
  keyboard shortcuts to complete them all:
  - [x] Opening a repository in a new window
  - [x] Opening the VSCode terminal
  - [x] Adding a new file
  - [x] Adding a new folder
  - [x] Deleting a file
  - [x] Deleting a folder
  - [x] Previewing a Markdown File
  - [x] Formatting a Markdown document
- [x] 🥚 **READMEs**: You can write a README file that describes the project you
      are working on, why it's helpful, and how someone can use it.
- 🥚 **NPM**: You can use NPM commands to verify your code's quality, this
  includes ...
  - [x] using `npm install` to install a project's dependencies
  - [x] reading a `package.json` file to find which scripts are available for
        the project
  - [x] use `npm run <script>` to execute an npm script
- 🥚 **Formatting Code**: You can use Prettier to make sure all the code in your
  project is well-formatted:
  - [x] You can use the Prettier VSCode extension to format a document while you
        are writing it
  - [x] You can use `npm run format` to format all of the documents in your
        project
  - [x] You can use `npm run format:check` to make sure all files are
        well-formatted
- 🥚 **Linting Code**: You can ...
  - [x] use `npm run lint:md` to check all Markdown files in your folder for
        linting mistakes
  - [x] you can fix all linting mistakes reported by `npm run lint:md`
- 🥚 **Spell Check**: You can ...
  - [x] use the Code Spell Checker VSCode extension to correct spelling mistakes
        in your Markdown documents while you are writing
  - [x] use `npm run spell-check` to check the spelling in all the files of your
        project
  - [x] update `.cspell.json` to add words that should be allowed in your
        project
- 🥚 **File and Folder Naming Conventions**: You can ...
  - [x] identify and follow the naming conventions for the project you are
        working on.
  - [x] use `npm run lint:ls` to check that all files and folders follow the
        project's naming conventions.
  - [x] you can correct any file and folder names to make them match the project
        conventions.
- [x] 🐣 **File Extensions**: You can identify all of the languages covered at
      HYF and give the correct file extension. You don't need to know the
      languages, just recognize them.
- [ ] 🐣 **Touch Typing**: You can write a Markdown file without looking at your
      keyboard to find any letters, numbers or special characters. (slowly is
      ok!)

### _**🥚 [1. Local Development With Git](./1-local-development-with-git)**_

Practice using Git to save and organize your development process. You will learn
how you can use Git to go back to previous versions of your project, and to work
on different changes in parallel.

- 🥚 **Git**: Using the CLI you can ...
  - [x] Initialize a new git repository with `git init`
  - [x] stage changes using `git add <path>`
  - [x] commit changes using `git commit -m <message>`
  - [x] display your repository's git history using `git log`
  - [x] create a new branch using `git branch <branch-name>`
  - [x] check out a branch using `git checkout <branch-name>`
  - [x] create a new branch and check it out using
        `git checkout -b <branch-name>`
  - [x] merge changes from one branch to another using `git merge <branch-name>`
  - [x] return to a previous version of your project with `git log` and
        `git checkout <commit-hash>`
  - [ ] stash and retrieve uncommitted changes with `git stash` and `git pop`
- 🥚 **`.gitignore`**: You can use a `.gitignore` file to describe which files
  you don't want included in your git history.
- 🥚 **VSCode**: You can ...
  - [ ] use the _Git Graph_ extension to to visualize your repo's commit history
  - [ ] use the _Git Lens_ extension to investigate your repository's commit
        history
  - [x] explain how the file tree in VSCode can show you which files have
        uncommitted changes
- [ ] 🥚 **Atomic Commits**: You can save your development progress using small
      commits with clear and helpful message.
- [ ] 🐣 **Feature Branches**: You can organize your development process using
      branches. You can create a new branch for each part of your project and
      merge those changes to `main` when they are finished.

### _**🥚 _[2. Local/Remote Development](./0-local-development-without-git)_**_
<hr/>

Learn how you can connect your local Git repositories with a GitHub repository
to add more structure to your development process and to share your projects.

- 🥚 **Licenses**: You can ...
  - [ ] explain why it's important to include a license for your code on GitHub
  - [ ] choose a license for your projects that matches how you want others to
        use your code
- [x] 🥚 **GitHub SSH Key**: You can connect your computer to your GitHub
      account using an SSH key, clone using the SSH link, and push/pull using
      your SSH connection.
- 🥚 **GitHub Repositories**: You can ...
  - [x] create new repository on GitHub
  - [x] write a description for your repository
  - [ ] turn on GitHub Pages
  - [x] configure your repository to block pushing to `main`
  - [x] configure your repository to block merging to `main` until Continuous
        Integration (CI) checks have passed

- 🥚 **Git Remote/Local Connection**: You can ...
  - [x] clone a remote repository to your computer
  - [x] initialize a new repository locally and connect it to an empty remote
        repository
  - [x] `push` and `pull` changes between remote & local branches
- [x] 🥚 **Pull Requests**: You can create a pull request between two branches
      in your repository and merge changes without causing any conflicts.
- [x] 🥚 **PR templates**: You can use a PR template to add a checklist to all
      of your PRs so you are sure the code is great before merging to `main`.
- [x] 🥚 **Continuous Integration**: You can use GitHub Actions to check your
      code's quality before merging a pull request to the `main` branch.
- [x] 🐣 **Local/Remote Branching Workflow**: You can use a branching workflow
      that keeps mistakes away from the `main` branch and prevents conflicts
      from happening in GitHub. For each contribution to the project you can ...
  1. Check out a new local branch and write your code
  2. Check out `main` on your local machine
  3. Pull changes from remote `main` to local `main`
  4. Merge changes locally from `main` to your new branch
     - Fix any conflicts on your new branch before pushing!
  5. Format and lint your code
  6. Push your new local branch to your remote repository
  7. Open a Pull Request from the new branch to `main`
     - Go through the PR's checklist to make sure everything is correct
     - Make sure all CI checks pass!
  8. Merge your new remote branch to `main`
     - You can delete the branch after it's merged if you want to

###  🐣 **_[3. Remote Collaboration](./3-remote-collaboration)_**
<hr/>

Learn how to collaborate with a group on a single project hosted in a GitHub
repository. Practice using GitHub's project management features to organize your
group's tasks and to double-check your project's code quality.

- [ ] 🥚 **Contributor Guidelines**: Your group can define contributor
      guidelines that describes how everyone can add their work to the project.
- [ ] 🥚 **Code of Conduct**: Your group can write a code of conduct for your
      project that describes how everyone should communicate and what behavior
      is considered unproductive.
- [ ] 🥚 **Repository Contributors**: You can add group members as collaborators
      in a repository.
- [ ] 🥚 **Issue Templates**: Your group can use issue templates to make sure
      all issues are complete and relevant to the project. in your repository
      and merge changes without causing any conflicts.
- [x] 🥚 **Code Review**: You can use the PR Template and CI checks to review
      another group member's code before merging it to `main`.
- 🐣 **GitHub Project Management**: You can ...
  - [ ] use issues to define tasks in a group project. Each issue should have a
        helpful title, complete description, and helpful labels.
  - [ ] use a Project Board to organize a project's issues
  - [ ] claim issues and track your progress with the project board
  - [ ] link a PR to your claimed issue
  - [ ] assign someone to review your PR
  - [ ] use GitHub code review features to discuss your code with your reviewer
- [x] 🐣 **Fixing Merge Conflicts**: You can fix merge conflicts by selecting
      which change to keep, or deciding how to combine them. If a conflict
      occurs in a GitHub PR you can pull both branches and fixh the conflict
      locally.

### _**🐥[4. Open Source Development](./4-open-source-development)**_
<hr/>

Explore the wider world of Open Source software by learning how communities of
independent developers write and maintain the code we all rely on.

- [ ] 🥚 **Finding Issues**: You can read through an open source project's
      issues and determine if there is one you can help with.
- [x] 🐣 **Commenting in Issues**: You can productively join the conversation in
      an open source projects' issues.
- [x] 🐥 **Opening Issues**: You can open a helpful issue in an open source
      repository, following their contributor guidelines and code of conduct.

- [ ] 🐔 **Contributing**: You can create a fork of an open source project and
       open a PR that contributes something helpful and follows the Contributor
      Guidelines and Code of Conduct.

**Comment by minjupgeorge** [comment]
### Week 2

## I Need Help With:
 Issues solved
        
## What went well?
1. clone the repository
2. add a file to student-bios in local machine
3. commit the changes
4. push branch to class repository
5. pull request has been initiated

## What went less well?
1. picture uploading

## Lessons Learned
1. while executing git commands first execute "git pull origin main"
2. To Organize the folders in the system
3. Need to commit the change before push command
 
## Sunday Prep Work
1. I have started reading workflow modules

**Comment by danielhalasz** [comment]
@minjupgeorge you are doing an excellent job in this module, it is so nice to see that!! do you still have some questions regarding image uploading?

**Comment by minjupgeorge** [comment]
@danielhalasz  Its getting clear. still am practicing to upload am image from my local machine and trying to do a pull request. last time I have used html tags for uploading an image froma folder in the repository .

**Comment by minjupgeorge** [comment]
@danielhalasz finally I have uploaded the image from my local m/c on the branch and then merged with repository main. 
now am clear with image uploading .

**Comment by danielhalasz** [comment]
@minjupgeorge hm, interesting approach. an easier way is to add the image locally to the student-bios/assets folder..and then just push to the team repo..after a review, it can be merged to the main branch

**Comment by minjupgeorge** [comment]
@danielhalasz yeah thats the best approch

**Comment by minjupgeorge** [comment]
### Week 2

## I Need Help With:
 Issues solved
        
## What went well?
1. Created the issue
2. Then moved the issue to the project board after assigning myself
3. Pulled the most recent code from master created a new branch for my md file

Writing the content:

added my intro file in /people according to group conventions
linked to my file from /README.md according to group conve

Before pushing my branch:

formatted the code
fixed all spelling mistakes and added new words to the dictionary
fixed all linting mistakes

After pushing my branch:

created a new PR linked to this issue
make sure all of the checks passed
assigned someone to review and merge my PR
1. clone the repository
2. add a file to student-bios in local machine
3. commit the changes
4. push branch to class repository
5. pull request has been initiated

## What went less well?
1. picture uploading
2. picture uploading issue resolved

## Lessons Learned
1. while executing git commands first execute "git pull origin main"
3. To Organize the folders in the system
4. Need to commit the change before push command
5. Learned how to resolve the conflict 

 
## Sunday Prep Work
1. I have started reading workflow modules

**Comment by colevandersWands** [comment]
> npx prettier --write

you can use the `npm run format` script in the repo to format all of the files, if you look in the `package.json` file you will see that it uses `prettier --write`.  

`npm run format:check` will check if the files are formatted but does not change the code in the files. When it asks "forgot to run prettier?" it just means that your files weren't formatted before checking.

try running the commands in this order and it should work
- `npm run format`
- `npm run format:check`

**Comment by minjupgeorge** [comment]
Hi @colevandersWands ...

Thanks alot the information.
I have executed the 'npm run format' and 'npm run format:check' . Now it's working.

**Comment by minjupgeorge** [comment]
Hi @danielhalasz,

I got error while trying put a table with 6 columns . 
![image](https://user-images.githubusercontent.com/32439811/159023164-645541d4-1bdf-4294-be52-491419318850.png)
Is there any specific rules about the space,length or no: of columns in a table ?

**Comment by minjupgeorge** [comment]
Hi @danielhalasz ,
I have created a branch 'minjutable' for creating table in README file. There I gt the error...after many trials...it gt uploaded

**Comment by danielhalasz** [comment]
@minjupgeorge if you check the error message, you can see that the markdown linting uses a configuration file called .markdownlint.jsonc and in this file, the MD013 setting is causing your error. you can either change the setting or the md file 🙂 

``` // MD013/line-length - Line length
  "MD013": {
    // Number of characters
    "line_length": 80,
    // Number of characters for headings
    "heading_line_length": 80,
    // Number of characters for code blocks
    "code_block_line_length": 80,
    // Include code blocks
    "code_blocks": false,
    // Include tables
    "tables": false,
    // Include headings
    "headings": true,
    // Include headings
    "headers": true,
    // Strict length checking
    "strict": false,
    // Stern length checking
    "stern": false
  },```

**Comment by minjupgeorge** [comment]
Thanks alot @danielhalasz . So better to change the settings .. how can I do it..!

**Comment by minjupgeorge** [comment]
Me 💻 🙏🏻 
---
**What was not clear, where did you get stuck?**
---
+ First I had difficulty with NPM run commands especially the  Prettier.Then through several false commits I have learned to run commands in proper order .Thanks alot @colevandersWands for helping me .

+ I got stuck when I had conflict issues by forgetting to command `get pull origin main` as the first step. Thank you @danielhalasz for correcting me for to upload image after editing in the local machine.
Every coaches are having lot of patience towards us.

**What was clear, what did you master?**
---
+ I have clear idea about  Git, Visual Studio Code (VSCode), the Command Line Interface (CLI) and NPM commands.
+ I have started enjoying to resolve the conflicts of my team members which helped me to gain more confidence.
+ My team members especially my team leader @TheRick88 and our coaches @danielhalasz , @prgrmfl   supported through out the correction phase of a authentication conflict by one of our team member .I have mastered in resolving it.
+ I have understand the how to create group project and GitHub Project management features.

**Where can you still use some help?**
---
How multiple team members can work on same file at same time?

**Where can you help others moving forwards?**
----
+ Being in a team member , I feel its my responsibility to help others issues and indirectly it helps me to have a good basic understanding of my knowledge about the module.
+ The Team-spirit which keep me motivated every time when I came across new errors. 
+ [x] Creating repo,cloning, push and pull git commands.
+ [x] Creating Tables and Inserting pictures
+ [x] Resolving Conflicts

The Course
<hr>
**What can there be more of?**
     
    Each of the Coaches doing a fantastic job.Hats off to each of you.
    It would be better if we have more videos of ' how each module works ' . @talmurshidi for uploading videos in the group.
   
**What can there be less of?**

I think we are having less time so everyone is rushing to complete and that can be solved by coming early so that doubts can be discussed and be ready for the classes.

**What material were most helpful (from HYF or elsewhere)?**

    Discord channel where we are having feel of working in a great organization .
    HYF StudyMaterials are more than enough.
    googling
**What HYF material was least helpful?**
   I didn;t feel any material which is least helpful

**Any suggestions for future classes?**

Every thing is well organized. still if possible please extend the time of group project work which we usually having in the end which is very important part of learning. Practice makes perfect :)

**Comment by danielhalasz** [comment]
@minjupgeorge thank you for the feedback, you can always come from 11 on Sundays and discuss material together. Was the time enough past Sunday to do the work in the smaller teams, or you would prefer to have even more?

**Comment by minjupgeorge** [comment]
Hi @danielhalasz Thank you for your comments and valuable inputs.
Yes,I will try to come as much as early  as possible. Past Sunday it was fine since only four in a group. we were able to create project board. assigned works and raised deliverable issue.
